src/Evaluate.ts

```diff
@@ -0,0 +1,598 @@
+
+  get updatedTimestamp(): number {
+    return this._updatedTimestamp;
+  }
+
+  set updatedTimestamp(value: number) {
+    this._updatedTimestamp = value;
+  }
+
+  public getToggle(key: string): Toggle | undefined {
+    return this._toggles[key];
+  }
+
+  public clear() {
+    this._toggles = {};
+    this._segments = {};
+  }
+}
+
+export class Toggle {
+  private _key: string;
+  private _enabled: boolean;
+  private _version: number;
+  private _forClient: boolean;
+  private _disabledServe: Serve;
+  private _defaultServe: Serve;
+  private _rules: Rule[];
+  private _variations: any[];
+
+  constructor(json: any) {
+    this._key = json.key;
+    this._enabled = json.enabled || false;
+    this._version = json.version || 1;
+    this._forClient = json.forClient || false;
+    this._disabledServe = new Serve(json.disabledServe);
+    this._defaultServe = new Serve(json.defaultServe);
+    this._rules = [];
+    for (const r of json.rules || []) {
+      this._rules.push(new Rule(r));
+    }
+    this._variations = json.variations || [];
+  }
+
+  get key(): string {
+    return this._key;
+  }
+
+  set key(value: string) {
+    this._key = value;
+  }
+
+  get enabled(): boolean {
+    return this._enabled;
+  }
+
+  set enabled(value: boolean) {
+    this._enabled = value;
+  }
+
+  get version(): number {
+    return this._version;
+  }
+
+  set version(value: number) {
+    this._version = value;
+  }
+
+  get forClient(): boolean {
+    return this._forClient;
+  }
+
+  set forClient(value: boolean) {
+    this._forClient = value;
+  }
+
+  get disabledServe(): Serve {
+    return this._disabledServe;
+  }
+
+  set disabledServe(value: Serve) {
+    this._disabledServe = value;
+  }
+
+  get defaultServe(): Serve {
+    return this._defaultServe;
+  }
+
+  set defaultServe(value: Serve) {
+    this._defaultServe = value;
+  }
+
+  get rules(): Rule[] {
+    return this._rules;
+  }
+
+  set rules(value: Rule[]) {
+    this._rules = value;
+  }
+
+  get variations(): any[] {
+    return this._variations;
+  }
+
+  set variations(value: any[]) {
+    this._variations = value;
+  }
+
+  public eval(user: FPUser, segments: { [key: string]: Segment }, defaultValue: any): FPToggleDetail {
+    if (!this._enabled) {
+      return this.disabledResult(user, this._key, defaultValue);
+    }
+    let warning: string | undefined;
+    if (this._rules?.length > 0) {
+      for (let i = 0; i < this._rules.length; i++) {
+        const rule = this._rules[i];
+        const hitResult = rule.hit(user, segments, this._key);
+        if (hitResult.hit) {
+          return this.hitValue(hitResult, defaultValue, i);
+        }
+        warning = hitResult.reason;
+      }
+    }
+    return this.defaultResult(user, this._key, defaultValue, warning);
+  }
+
+  private hitValue(hitResult: IHitResult | undefined, defaultValue: any, ruleIndex?: number): FPToggleDetail {
+    const res = {
+      value: defaultValue,
+      ruleIndex: ruleIndex || null,
```

Comment:
if ruleIndex is 0, expression 0 || null will return null, actually we need 0 not null.
same with line 187